pkg/queue/breaker.go

```diff
@@ -76,83 +90,105 @@ func (b *Breaker) Maybe(thunk func()) bool {
+	if size > 0 {
+		return b.sem.AddCapacity(size)
+	} else {
+		return b.sem.ReduceCapacity(-size)
+	}
+}
+
+// Capacity retrieves the capacity of the breaker.
+func (b *Breaker) Capacity() int32 {
+	return b.sem.capacity
+}
+
+// NewSemaphore creates a semaphore with the desired maximal and initial capacity.
+func NewSemaphore(maxCapacity, initialCapacity int32, logger *zap.SugaredLogger) *Semaphore {
 	if initialCapacity < 0 || initialCapacity > maxCapacity {
 		panic(fmt.Sprintf("Initial capacity must be between 0 and maximal capacity. Got %v.", initialCapacity))
 	}
 	queue := make(chan token, maxCapacity)
-	sem := Semaphore{queue: queue}
+	sem := Semaphore{queue: queue, logger: logger}
 	if initialCapacity > 0 {
 		sem.AddCapacity(initialCapacity)
 	}
 	return &sem
 }
 
-// Semaphore is an implementation of a semaphore based on Go channels
-// The number of available tokens is the number of elements in the buffered channel
+// Semaphore is an implementation of a semaphore based on Go channels.
+// The presence of elements in the `queue` buffered channel correspond to available tokens.
+// Hence the max number of tokens to hand out equals to the size of the channel.
+// `capacity` defines the current number of tokens in the rotation.
 type Semaphore struct {
 	queue    chan token
 	token    token
 	reducers int32
 	capacity int32
+	logger   *zap.SugaredLogger
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
 func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+			// this should never happen
+			s.logger.Error("Semaphore release error: returned tokens must be <= acquired tokens")
+		}
+	}
 }
 
-// ReduceCapacity removes tokens from the rotation
-// It tries to acquire as many tokens as possible, if there are not enough tokens in the queue,
-// it postpones the operation for the future by increasing the `reducers` counter
-func (s *Semaphore) ReduceCapacity(size int32) error {
+// AddCapacity increases the number of tokens in the rotation.
+// Increases only if the size is positive, does nothing otherwise.
+// An error is returned if not all capacity could be added to the semaphore.
+// This error could happen when the number of tokens exceeds the semaphore's queue size.
+func (s *Semaphore) AddCapacity(size int32) error {
```

Comment:
Should this take the reducers into account as well? I.e.: Substract the amount of reducers from the amount of capacity to be added instead of only letting `Release` handle that?

Reply:
If I understood you correctly, we don't need to do that in `AddCapacity`, since the method will be called only if new endpoints are added. If the endpoints are deleted, then the `ReduceCapacity` should handle that.

Reply:
Well, you refer to `AddCapacity` for both cases - releasing the tokens that were already taken [1] and adding the new ones [2], like it was before, right?
One possible problem here is that we increase the capacity even in [1] - which is not what we want, the capacity should be only updated for [2], when new endpoints are added. Otherwise it would constantly increase when tokens are released leading to the case when more tokens are in rotation then the number of available endpoints. Thus I split both use cases into `AddCapacity` and `Release` methods. On a side note - I find it more readable to separate the concerns.

Reply:
No I'm not referring to `AddCapacity` for both cases like before. The thing is, when there are still reducers in flight, you might as well substract them from the to-be-added capacity before adding any capacity.

For each reducer, you want to take a token out of the rotation. So in this case, you'd just not create a new token for each reducer.

Reply:
As discussed it per slack as well. In our particular use case (semaphore tokens correspond to the number of available endpoints), it is rather improbable that AddCapacity would be called right after ReduceCapacity. Thus the impact of the change would not bring the value right away. So we agreed to leave the method as it is for now, unless we find it being a bottleneck in feature.